ftw/tokenauth/oauth2/browser/oauth2_token.py

```diff
@@ -247,11 +254,36 @@ def validate_jwt_grant(self, assertion, service_key):
         # https://tools.ietf.org/html/rfc7523#section-3
         processor = JWTBearerGrantProcessor(self.plugin.get_token_uri())
         try:
-            processor.verify(assertion, service_key)
+            return processor.verify(assertion, service_key)
         except (VerificationError, InvalidTokenError) as exc:
             raise InvalidGrant(str(exc))
 
-    def issue_access_token(self, service_key):
+    def verified_subject(self, claimset, service_key):
+        """Verify the claim's subject and return it.
+
+        If the subject does not match the userid of the service_key, the
+        'Impersonate user' permission is required.
+        """
+        subject = claimset['sub']
+        actor = service_key['user_id']
+        if subject != actor:
+            # Check if actor is allowed to impersonate
+            uf = getToolByName(self.context, 'acl_users')
+            user = uf.getUserById(actor)
+            if not user:
+                raise(InvalidGrant('Service key user not found.'))
+            user = user.__of__(uf)
+            old_security_manager = getSecurityManager()
+            newSecurityManager(self.request, user)
+            allow_impersonation = getSecurityManager().checkPermission(
+                ImpersonateUser, self.context)
+            setSecurityManager(old_security_manager)
+            if not allow_impersonation:
+                raise InvalidGrant(str(SubjectMismatch(
```

Comment:
Is `str(SubjectMismatch(` necessary here?
Usually `str(Exception(x)) == x`, but I don't know the internals of `SubjectMismatch`.. 😉

Reply:
I've removed the `SubjectMismatch` exception, as it's no longer needed.